packages/common/src/converter/payload-converter.ts

```diff
@@ -140,3 +149,75 @@ export class BinaryPayloadConverter extends AsyncFacadePayloadConverter {
     return content.data as any;
   }
 }
+
+/**
+ * Converts between protobufjs Message instances and serialized Protobuf Payload
+ */
+export class ProtobufPayloadConverter extends AsyncFacadePayloadConverter {
+  public encodingType = encodingTypes.METADATA_ENCODING_PROTOBUF;
+
+  constructor(protected protobufClasses?: Record<string, unknown>) {
+    super();
+
+    if (protobufClasses && typeof protobufClasses !== 'object') {
+      throw new DataConverterError('protobufClasses must be an object');
+    }
+  }
+
+  public toDataSync(value: unknown): Payload | undefined {
+    const isProtobufMessageInstance =
+      this.protobufClasses && value && typeof value === 'object' && this.protobufClasses[value.constructor.name];
+    if (!isProtobufMessageInstance) {
+      return undefined;
+    }
+
+    // Safe to non-null assert due to above check
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const messageClass = this.validateMessageClass(this.protobufClasses![value.constructor.name]);
+
+    return {
+      metadata: {
+        [METADATA_ENCODING_KEY]: encodingKeys.METADATA_ENCODING_PROTOBUF,
+        [METADATA_MESSAGE_TYPE_KEY]: u8(value.constructor.name),
+      },
+      data: messageClass.encode(value).finish(),
+    };
+  }
+
+  public fromDataSync<T>(content: Payload): T {
+    if (content.data === undefined || content.data === null) {
+      throw new ValueError('Got payload with no data');
+    }
+    if (!content.metadata || !(METADATA_MESSAGE_TYPE_KEY in content.metadata)) {
+      throw new ValueError(`Got protobuf payload without metadata.${METADATA_MESSAGE_TYPE_KEY}`);
+    }
+    if (!this.protobufClasses) {
+      throw new DataConverterError(
+        'Unable to deserialize protobuf message without protobufClasses provided to DefaultDataConverter'
+      );
+    }
+
+    const messageClassName = str(content.metadata[METADATA_MESSAGE_TYPE_KEY]);
+    const messageClass = this.protobufClasses[messageClassName];
+    if (!messageClass) {
+      throw new DataConverterError(
+        `Got a \`${messageClassName}\` protobuf message but cannot find corresponding message class in protobufClasses`
+      );
+    }
+
+    return this.validateMessageClass(messageClass).decode<T>(content.data);
```

Comment:
This class should be validated during construction

Reply:
Do you mean going through `protobufClasses` in the constructor and collecting all the values that pass validation?

Reply:
Like this: ![image](https://user-images.githubusercontent.com/251288/146605005-4fd32fd9-6935-42bd-935b-01202a50b508.png)